src/litdata/utilities/shuffle.py

```diff
@@ -84,83 +85,144 @@ def _associate_chunks_and_internals_to_ranks(
+                        num_shared_workers_for_this_rank -= 1
+
+                    # We consumed entirely the chunk of the worker we were tracking, let's break
+                    # TODO: Maybe, we can prevent loading over and over for each worker
+                    if num_shared_workers_for_this_rank == 0 and current_worker_chunk_index == chunk_index:
+                        if chunk_index not in max_trackers:
+                            max_trackers[chunk_index] = (
+                                local_rank * num_workers + worker_tracker_idx % num_workers,
+                                counter,
+                            )
+                        else:
+                            if max_trackers[chunk_index][1] < counter:
+                                max_trackers[chunk_index] = (
+                                    local_rank * num_workers + worker_tracker_idx % num_workers,
+                                    counter,
+                                )
+                        break
 
-            counter = 0
-            worker_idx = 0  # reset the worker_idx
-            while True:
-                current_chunks = workers_chunks[worker_idx]
-                current_intervals = workers_intervals[worker_idx]
+                    if num_samples_left_for_this_worker_chunk != batch_size:
+                        num_of_samples_to_carry_to_next_chunk = batch_size - num_samples_left_for_this_worker_chunk
 
-                if len(current_intervals) == 0:
-                    break
+                    if remover != batch_size:
+                        num_of_samples_to_carry_to_next_chunk = None
 
-                if current_intervals[0] > batch_size:
-                    current_intervals[0] -= batch_size
-                    counter += batch_size
-                    worker_idx = (worker_idx + 1) % num_workers
-                else:
-                    counter += current_intervals[0]
-                    current_intervals.pop(0)
-                    current_chunk = current_chunks.pop(0)
-                    worker_idx = (worker_idx + 1) % num_workers
+                if num_of_samples_to_carry_to_next_chunk is None:
+                    worker_tracker_idx += 1
 
-                    if current_chunk == shared_chunk:
-                        break
+    to_disable = {}
+    for chunk_index, worker_ids in shared_chunks.items():
+        last_worker_idx = max_trackers[chunk_index][0]
+        to_disable[chunk_index] = [worker_idx for worker_idx in worker_ids if worker_idx != last_worker_idx]
+    return to_disable
 
-            counters.append(counter)
 
-        max_counter = np.argmax(counters)
-        disable_ranks = [rank for rank in ranks if rank != ranks[max_counter]]
-        for rank in disable_ranks:
-            if rank not in disable_deletion_ranks:
-                disable_deletion_ranks[rank] = [shared_chunk]
+def _get_shared_chunks(workers_chunks: List[List[int]]) -> Dict[int, List[int]]:
```

Comment:
Let's add some comments in those functions. This would be helpful for any reader.